# Dataset Generation

In [1]:
from  google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd

# Let's load the original tsv file

In [ ]:
data = pd.read_csv('reviews.tsv', sep='\t', error_bad_lines=False)
# Randomly selecting 100k reviews
balanced_data = data.sample(n=100000, random_state=1)
# Saving the sampled dataset to a new TSV file
balanced_data.to_csv('sampled_dataset.tsv', sep='\t', index=False)

C:\Users\Sam\AppData\Local\Temp\ipykernel_9692\232109533.py:1: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  data = pd.read_csv('reviews.tsv', sep='\t', error_bad_lines=False)
Skipping line 20773: expected 15 fields, saw 22
Skipping line 39834: expected 15 fields, saw 22
Skipping line 52957: expected 15 fields, saw 22
Skipping line 54540: expected 15 fields, saw 22

Skipping line 80276: expected 15 fields, saw 22
Skipping line 96168: expected 15 fields, saw 22
Skipping line 96866: expected 15 fields, saw 22
Skipping line 98175: expected 15 fields, saw 22
Skipping line 112539: expected 15 fields, saw 22
Skipping line 119377: expected 15 fields, saw 22
Skipping line 120065: expected 15 fields, saw 22
Skipping line 124703: expected 15 fields, saw 22

Skipping line 134024: expected 15 fields, saw 22
Skipping line 153938: expected 15 fields, saw 22
Skipping line 156225: expected 15 fields, saw 22


In [ ]:
data.shape,balanced_data.shape

((2640254, 15), (100000, 15))

# Reading balanced Data

In [ ]:
balanced_data=pd.read_csv('sampled_dataset.tsv', sep='\t', error_bad_lines=False)
balanced_data.head(5)

C:\Users\Sam\AppData\Local\Temp\ipykernel_9692\2756616047.py:1: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  balanced_data=pd.read_csv('sampled_dataset.tsv', sep='\t', error_bad_lines=False)


marketplace  customer_id       review_id  product_id  product_parent  \
0          US     48006260   RTCLHUE5XW7AS  B0083J78GY       955804173   
1          US     15333704   RZHL9W1IW3NL9  B00DF9YOIC       473310322   
2          US     42986671   R910MA53M5SOX  B001QVXHR0       768255570   
3          US     48229683  R32SNFLPNU71V9  B00A73PCT4       161179273   
4          US     14820508  R2BE84Z6ZM4YX6  B00007BG8P       847877046   

                                       product_title product_category  \
0       Canopy 2-Year Office Product Protection Plan  Office Products   
1  Silver / Gold Blue Butterfly 2 Year 2017-18 & ...  Office Products   
2  Printronic Remanufactured Ink Cartridge Replac...  Office Products   
3  2.4G Wireless Automatic USB Laser Handheld Bar...  Office Products   
4             Sony SVM-75LS Ink and Paper Value Pack  Office Products   

   star_rating  helpful_votes  total_votes vine verified_purchase  \
0          5.0            0.0          0.0    N                 Y   
1          5.0            0.0          0.0    N                 Y   
2          1.0            0.0          0.0    N                 Y   
3          1.0            0.0          0.0    N                 Y   
4          4.0            0.0          0.0    N                 Y   

                       review_headline  \
0                           Five Stars   
1          MUCH BETTER THAN I EXPECTED   
2                                  Ink   
3  bar code scanner not a good product   
4                      Clarity & Speed   

                                         review_body review_date  
0                                              Great  2015-02-28  
1  The pocket calendar is better than I expected,...  2013-12-14  
2  The cartridges are no good, because the ink do...  2013-04-04  
3  the scanner can not be set up for a carriage c...  2013-08-03  
4  I have been using the Sony dye-sub printers fo...  2010-01-21

# Preparing the Dataset and Displaying the Initial Data

In [ ]:
#Creating a new DataFrame with only 'Reviews' and 'Ratings' columns
dataset = balanced_data[['review_headline', 'review_body', 'star_rating']].copy()
#Concatenate 'review_headline' and 'review_body' to create a 'Reviews' column
dataset['Reviews'] = dataset['review_headline'] + ' ' + dataset['review_body']
#Renaming the star_rating column to Ratings
dataset.rename(columns={'star_rating': 'Ratings'}, inplace=True)

# Dropping the 'review_headline' and 'review_body' columns
dataset.drop(['review_headline', 'review_body'], axis=1, inplace=True)
dataset = dataset[['Reviews', 'Ratings']]
dataset.head(10)

Reviews  Ratings
0                                   Five Stars Great      5.0
1  MUCH BETTER THAN I EXPECTED The pocket calenda...      5.0
2  Ink The cartridges are no good, because the in...      1.0
3  bar code scanner not a good product the scanne...      1.0
4  Clarity & Speed I have been using the Sony dye...      4.0
5  Good I'm still trying to become fimilar with i...      3.0
6  Brother Compatible Cartridges Just installed c...      5.0
7                          Five Stars Great markers!      5.0
8            One Star it only lasted for a few pages      1.0
9  Great for Showcasing Your Marketing Pieces As ...      5.0

In [5]:
dataset = pd.read_csv('/content/drive/My Drive/Upwork/dataset.csv')
dataset.drop(['Unnamed: 0'], axis=1, inplace=True)

In [6]:
dataset.Reviews[6]

'Brother Compatible Cartridges Just installed cartridges - was worried as the shape seemed different.  Have had NO problems with the ink or printing thus far.  SO MUCH CHEAPER!!!! Do you make them for other printers?!!'

In [7]:
dataset.isna().sum()

Reviews    6
Ratings    0
dtype: int64

# Dropping NAN values

In [8]:
dataset.dropna(inplace=True)

In [9]:
dataset.isna().sum()

Reviews    0
Ratings    0
dtype: int64

# Count of Reviews by Star Ratings

In [10]:
dataset['Ratings'].value_counts()

5.0    59848
4.0    15775
1.0    11830
3.0     7287
2.0     5254
Name: Ratings, dtype: int64

# Classification of reviews by Stars

In [11]:
positive_reviews = dataset[dataset['Ratings'].isin([4, 5])]
negative_reviews = dataset[dataset['Ratings'].isin([1, 2])]
neutral_reviews = dataset[dataset['Ratings'] == 3]

In [12]:
print('Number of positive reviews')
positive_reviews.count()

Number of positive reviews


Reviews    75623
Ratings    75623
dtype: int64

In [13]:
print('Number of Negative reviews')
negative_reviews.count()

Number of Negative reviews


Reviews    17084
Ratings    17084
dtype: int64

In [14]:
print('Number of Negative reviews')
neutral_reviews.count()

Number of Negative reviews


Reviews    7287
Ratings    7287
dtype: int64

# Word Embedding

# word2vec-google-news-300

The "word2vec-google-news-300" model is a popular and widely-used pre-trained Word2Vec model. It is trained on a large Google News dataset and provides 300-dimensional word vectors, making it a valuable resource for a wide range of NLP applications

In [15]:
import gensim.downloader as api

In [16]:
wv=api.load('word2vec-google-news-300')

[==================================================] 100.0% 1662.8/1662.8MB downloaded


In [17]:
similarity=wv.similarity('cartridges', 'product')
similarity

0.2240968

In [18]:
similarity=wv.similarity(w1='teacher',w2='student')
print('similarity between student and teacher is :',similarity)

similarity between student and teacher is : 0.63013655


In [19]:
analogy_result=wv.most_similar(positive=['father', 'woman'], negative=['man'], topn=1)
analogy_result

[('mother', 0.8462507128715515)]

In [20]:
from gensim.models import Word2Vec

# Extracting the "Reviews" column
reviews = dataset['Reviews']

# Splitting the reviews into lists of words (no preprocessing)
Reviews = [review.split() for review in reviews]

# Training the Word2Vec model
model = Word2Vec(Reviews, vector_size=300, window=13, min_count=9)

# Save the model
model.save("reviews_word2vec.model")

In [21]:
similarity = model.wv.similarity('cartridges', 'product')

print(f"Similarity using custom Word2Vec model: {similarity}")


Similarity using custom Word2Vec model: 0.21525433659553528


In [22]:
similarity = model.wv.similarity('excellent', 'product')

print(f"Similarity using custom Word2Vec model: {similarity}")


Similarity using custom Word2Vec model: 0.11104869842529297


# Example 1

In [23]:
analogy_result=model.wv.most_similar(positive=['brother','mother'],negative=['sister'],topn=1)
analogy_result

[('Dell', 0.6989747285842896)]

# Example 2

In [24]:
analogy_result = model.wv.most_similar(positive=['Sony', 'quality'], negative=['best'], topn=1)
print("Analogy Result:", analogy_result)

Analogy Result: [('quality,', 0.5170270800590515)]


# Example 3

In [25]:
analogy_result =model.wv.most_similar(positive=['Cartridges', 'printers'], negative=['product'], topn=1)
print("Analogy Result:", analogy_result)

Analogy Result: [('printers,', 0.6918996572494507)]


# Conclusion

In terms of encoding semantic similarities between words, the pretrained model ("word2vec-google-news-300") beats the custom-trained model. This is due to the pretrained model's utilization of a large and diverse dataset during training, which allows it to learn larger language connection and contexts, resulting in greater similarity scores.

# Simple Models

In [26]:
import gensim
from gensim.models import Word2Vec
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import Perceptron
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# Load the pre-trained Word2Vec model
word2vec_model = Word2Vec.load("reviews_word2vec.model")

# Split your dataset into training and testing sets
from sklearn.model_selection import train_test_split
X = dataset['Reviews']  # Extract the 'reviews' column as features
y = dataset['Ratings']  # Extract the 'ratings' column as labels

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [27]:
# Calculating average Word2Vec vectors for each review in the dataset
def calculate_average_word2vec(review, model):
    words = review.split()  # Assuming each review is a space-separated string
    vectors = []
    for word in words:
        if word in model.wv:  # Check if the word is in the Word2Vec model's vocabulary
            vectors.append(model.wv[word])
    if vectors:
        return np.mean(vectors, axis=0)
    else:
        return np.zeros(model.vector_size)

In [28]:
X_train_word2vec = np.array([calculate_average_word2vec(review, word2vec_model) for review in X_train])
X_test_word2vec = np.array([calculate_average_word2vec(review, word2vec_model) for review in X_test])

# Perceptron models

In [29]:
# Training Perceptron models
perceptron_word2vec = Perceptron()
perceptron_word2vec.fit(X_train_word2vec, y_train)
y_pred_word2vec = perceptron_word2vec.predict(X_test_word2vec)
accuracy_word2vec = accuracy_score(y_test, y_pred_word2vec)
print('accuracy with perceptron Model:',accuracy_word2vec)

accuracy with perceptron Model: 0.6779338966948347


In [30]:
from sklearn.metrics import classification_report
# Classification report for Perceptron with Word2Vec
print("Classification Report for Perceptron with Word2Vec:")
print(classification_report(y_test, y_pred_word2vec))

Classification Report for Perceptron with Word2Vec:
              precision    recall  f1-score   support

         1.0       0.67      0.57      0.62      2375
         2.0       0.28      0.09      0.14      1081
         3.0       0.35      0.25      0.29      1507
         4.0       0.35      0.36      0.35      3087
         5.0       0.79      0.89      0.84     11949

    accuracy                           0.68     19999
   macro avg       0.49      0.43      0.45     19999
weighted avg       0.65      0.68      0.66     19999



# SVM models

In [31]:
# Training SVM models
svm_word2vec = SVC(kernel='linear')
svm_word2vec.fit(X_train_word2vec, y_train)
y_pred_svm_word2vec = svm_word2vec.predict(X_test_word2vec)
accuracy_svm_word2vec = accuracy_score(y_test, y_pred_svm_word2vec)

In [32]:
accuracy_svm_word2vec

0.71943597179859

In [33]:
# Classification report for SVM with Word2Vec
print("Classification Report for SVM with Word2Vec:")
print(classification_report(y_test, y_pred_svm_word2vec))

Classification Report for SVM with Word2Vec:
              precision    recall  f1-score   support

         1.0       0.63      0.78      0.70      2375
         2.0       0.42      0.10      0.16      1081
         3.0       0.44      0.28      0.34      1507
         4.0       0.74      0.15      0.25      3087
         5.0       0.76      0.96      0.85     11949

    accuracy                           0.72     19999
   macro avg       0.60      0.46      0.46     19999
weighted avg       0.70      0.72      0.66     19999



# TF-IDF Feature Extraction

In [34]:
# TF-IDF Feature Extraction
tfidf_vectorizer = TfidfVectorizer()
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

In [35]:
# Training Perceptron model with TF-IDF features
perceptron_tfidf = Perceptron()
perceptron_tfidf.fit(X_train_tfidf, y_train)
y_pred_tfidf = perceptron_tfidf.predict(X_test_tfidf)
accuracy_tfidf_perceptron = accuracy_score(y_test, y_pred_tfidf)

In [36]:
# Classification report for Perceptron with TF-IDF
print("Classification Report for Perceptron with TF-IDF:")
print(classification_report(y_test, y_pred_tfidf))


Classification Report for Perceptron with TF-IDF:
              precision    recall  f1-score   support

         1.0       0.64      0.69      0.66      2375
         2.0       0.27      0.27      0.27      1081
         3.0       0.32      0.38      0.35      1507
         4.0       0.46      0.37      0.41      3087
         5.0       0.84      0.85      0.85     11949

    accuracy                           0.69     19999
   macro avg       0.51      0.51      0.51     19999
weighted avg       0.69      0.69      0.69     19999



In [37]:
# Training SVM model with TF-IDF features
svm_tfidf = SVC(kernel='linear')
svm_tfidf.fit(X_train_tfidf, y_train)
y_pred_svm_tfidf = svm_tfidf.predict(X_test_tfidf)
accuracy_tfidf_svm = accuracy_score(y_test, y_pred_svm_tfidf)

In [38]:
# Classification report for SVM with TF-IDF
print("Classification Report for SVM with TF-IDF:")
print(classification_report(y_test, y_pred_svm_tfidf))


Classification Report for SVM with TF-IDF:
              precision    recall  f1-score   support

         1.0       0.69      0.81      0.75      2375
         2.0       0.46      0.22      0.30      1081
         3.0       0.51      0.36      0.42      1507
         4.0       0.59      0.36      0.45      3087
         5.0       0.83      0.95      0.88     11949

    accuracy                           0.76     19999
   macro avg       0.61      0.54      0.56     19999
weighted avg       0.73      0.76      0.73     19999



# Comparison

In [39]:
# Comparing performances
print("Perceptron Word2Vec Accuracy:", accuracy_word2vec)
print("Perceptron TF-IDF Accuracy:", accuracy_tfidf_perceptron)

print("SVM Word2Vec Accuracy:", accuracy_svm_word2vec)
print("SVM TF-IDF Accuracy:", accuracy_tfidf_svm)

Perceptron Word2Vec Accuracy: 0.6779338966948347
Perceptron TF-IDF Accuracy: 0.6912845642282114
SVM Word2Vec Accuracy: 0.71943597179859
SVM TF-IDF Accuracy: 0.7579878993949698


# Conclusion

In summary, the results indicate that TF-IDF features coupled with the SVM model consistently outperform Word2Vec features in both Perceptron and SVM models. The SVM with TF-IDF achieves the highest accuracy, suggesting that this combination is the most effective for the classification task

## Feedforward Neural Networks (FNN)

A Feedforward Neural Network (FNN) is a fundamental type of artificial neural network where information flows unidirectionally, from the input layer through hidden layers (if any) to the output layer.

## a) Average Word2Vec Vectors as Input Features

In [59]:
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.metrics import accuracy_score

# Define the Feedforward Neural Network model
class FFN(nn.Module):
    def __init__(self, input_size, hidden_size1, hidden_size2, output_size):
        super(FFN, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size1)
        self.fc2 = nn.Linear(hidden_size1, hidden_size2)
        self.fc3 = nn.Linear(hidden_size2, output_size)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.fc3(x)
        return x



In [60]:

# Convert Word2Vec vectors to PyTorch tensors
X_train_word2vec_tensor = torch.tensor(X_train_word2vec, dtype=torch.float32)
X_test_word2vec_tensor = torch.tensor(X_test_word2vec, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train.values - 1, dtype=torch.long)  # Subtract 1 to start labels from 0
y_test_tensor = torch.tensor(y_test.values - 1, dtype=torch.long)    # Subtract 1 to start labels from 0



In [61]:
# Instantiate the model and set hyperparameters
input_size = 300  # Size of Word2Vec vectors
hidden_size1 = 50
hidden_size2 = 5
output_size = 5  #  there are 5 classes
model_avg = FFN(input_size, hidden_size1, hidden_size2, output_size)

# Define loss function and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model_avg.parameters(), lr=0.001)



In [62]:
# Training the model
num_epochs = 10
for epoch in range(num_epochs):
    optimizer.zero_grad()
    outputs = model_avg(X_train_word2vec_tensor)
    loss = criterion(outputs, y_train_tensor)
    loss.backward()
    optimizer.step()

# Test the model
with torch.no_grad():
    test_outputs = model_avg(X_test_word2vec_tensor)
    _, predicted = torch.max(test_outputs, 1)
    ffn_accuracy_avg = accuracy_score(y_test.values, predicted)
    print("Accuracy of FNN with average Word2Vec vectors: {:.2f}%".format(ffn_accuracy_avg * 100))

Accuracy of FNN with average Word2Vec vectors: 15.44%


## b) First 10 Word2Vec Vectors Concatenated as Input Features

In [77]:
Ratings = dataset['Ratings']
Ratings

0        5.0
1        5.0
2        1.0
3        1.0
4        4.0
        ... 
99995    5.0
99996    5.0
99997    3.0
99998    5.0
99999    5.0
Name: Ratings, Length: 99994, dtype: float64

In [84]:
# Modify X_train_word2vec and X_test_word2vec to keep only the first 10 vectors for each review
X_train_word2vec_first10 = np.array([review[:10] for review in X_train_word2vec])
X_test_word2vec_first10 = np.array([review[:10] for review in X_test_word2vec])

# Convert the modified Word2Vec vectors to PyTorch tensors
X_train_word2vec_first10_tensor = torch.tensor(X_train_word2vec_first10, dtype=torch.float32)
X_test_word2vec_first10_tensor = torch.tensor(X_test_word2vec_first10, dtype=torch.float32)




In [72]:
print(X_train_word2vec_first10_tensor.shape)
print(X_test_word2vec_first10_tensor.shape)

#X_train_word2vec_first10_tensor has a shape of torch.Size([79995, 10]), indicating 79,995 samples, each with 10 vectors.
#X_test_word2vec_first10_tensor has a shape of torch.Size([19999, 10]), indicating 19,999 samples, each with 10 vectors.

torch.Size([79995, 10])
torch.Size([19999, 10])


In [108]:
# Modify y_train to keep only the first 10 labels for each review
y_train_first10 = y_train[:79995]
y_test_first10 = y_test[:19999]

# Convert the modified labels to PyTorch tensors
y_train_first10_tensor = torch.tensor(y_train_first10.values - 1, dtype=torch.long)  # Subtract 1 to start labels from 0
y_test_first10_tensor = torch.tensor(y_test_first10.values - 1, dtype=torch.long)  # Subtract 1 to start labels from 0


In [109]:
# Instantiate a new FFN model for this task with correct input size
model_first10 = FFN(input_size=10, hidden_size1=50, hidden_size2=5, output_size=5)


# Define loss function and optimizer
optimizer_first10 = optim.Adam(model_first10.parameters(), lr=0.001)




In [110]:
import torch.nn.functional as F
num_classes = 5  # Number of classes in your classification task
y_train_one_hot = F.one_hot(y_train_first10_tensor, num_classes)



In [111]:
print(outputs_first10.shape)
print(y_train_one_hot.shape)
print(y_train_first10_tensor.shape)


torch.Size([79995, 5])
torch.Size([79995, 5])
torch.Size([79995])


In [112]:
# Training the model for the reduced dataset
for epoch in range(num_epochs):
    optimizer_first10.zero_grad()
    outputs_first10 = model_first10(X_train_word2vec_first10_tensor)
    loss_first10 = criterion(outputs_first10, y_train_first10_tensor)  # Use the original class labels, not one-hot encoded vectors
    loss_first10.backward()
    optimizer_first10.step()

# Test the model
with torch.no_grad():
    test_outputs_first10 = model_first10(X_test_word2vec_first10_tensor)
    _, predicted_first10 = torch.max(test_outputs_first10, 1)
    ffn_accuracy_first10 = accuracy_score(y_test_first10.values, predicted_first10)
    print("Accuracy of FNN with first 10 Word2Vec vectors: {:.2f}%".format(ffn_accuracy_first10 * 100))


Accuracy of FNN with first 10 Word2Vec vectors: 15.09%


# Conclusion

 Both Feedforward Neural Network models, trained on average Word2Vec vectors and the first 10 vectors, exhibited low accuracy around 15%. To improve, focusing on data cleaning, expanding the dataset, experimenting with diverse model architectures, optimizing hyperparameters, and exploring advanced embeddings like BERT could significantly enhance predictive capabilities. Addressing these areas is crucial for building a more accurate and reliable model.

## Recurrent Neural Networks (RNN)

## a) Simple RNN

In [149]:
# Define the SimpleRNN model
class SimpleRNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(SimpleRNN, self).__init__()
        self.rnn = nn.RNN(input_size, hidden_size, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        rnn_out, _ = self.rnn(x)
        # Extract the last time step's hidden state
        last_hidden_state = rnn_out[:, -1, :]
        # Use the last time step's hidden state for classification
        output = self.fc(last_hidden_state)
        return output
# Convert the first 10 Word2Vec vectors to PyTorch tensors
X_train_word2vec_first10_tensor = torch.tensor(X_train_word2vec_first10, dtype=torch.float32)
X_test_word2vec_first10_tensor = torch.tensor(X_test_word2vec_first10, dtype=torch.float32)

# Ensure that the input tensors have the correct shape [batch_size, seq_len, input_size]
X_train_word2vec_first10_tensor = X_train_word2vec_first10_tensor.unsqueeze(1).expand(-1, 10, -1)
X_test_word2vec_first10_tensor = X_test_word2vec_first10_tensor.unsqueeze(1).expand(-1, 10, -1)

# Instantiate the SimpleRNN model
rnn_model = SimpleRNN(input_size=10, hidden_size=10, output_size=5)

# Define loss function and optimizer
criterion_rnn = nn.CrossEntropyLoss()
optimizer_rnn = optim.Adam(rnn_model.parameters(), lr=0.001)

# Training the model
num_epochs = 10
for epoch in range(num_epochs):
    optimizer_rnn.zero_grad()
    rnn_outputs = rnn_model(X_train_word2vec_first10_tensor)
    loss_rnn = criterion_rnn(rnn_outputs, y_train_tensor)
    loss_rnn.backward()
    optimizer_rnn.step()

# Test the model
with torch.no_grad():
    test_rnn_outputs = rnn_model(X_test_word2vec_first10_tensor)
    _, predicted_rnn = torch.max(test_rnn_outputs, 1)
    rnn_accuracy = accuracy_score(y_test.values, predicted_rnn)
    print("Accuracy of Simple RNN: {:.2f}%".format(rnn_accuracy * 100))




Accuracy of Simple RNN: 8.17%


## b) Gated Recurrent Unit (GRU)


In [150]:
# Define the GRU model
class GRUModel(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(GRUModel, self).__init__()
        self.gru = nn.GRU(input_size, hidden_size, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        gru_out, _ = self.gru(x)
        output = self.fc(gru_out[:, -1, :])  # Use the last time step's output for classification
        return output




In [153]:
# Instantiate the GRU model
gru_model = GRUModel(input_size=10, hidden_size=10, output_size=5)

# Define loss function and optimizer for GRU
optimizer_gru = optim.Adam(gru_model.parameters(), lr=0.001)



In [154]:
# Training the GRU model
for epoch in range(num_epochs):
    optimizer_gru.zero_grad()
    gru_outputs = gru_model(X_train_word2vec_first10_tensor)
    loss_gru = criterion(gru_outputs, y_train_tensor)
    loss_gru.backward()
    optimizer_gru.step()

# Test the GRU model
with torch.no_grad():
    test_gru_outputs = gru_model(X_test_word2vec_first10_tensor)
    _, predicted_gru = torch.max(test_gru_outputs, 1)
    gru_accuracy = accuracy_score(y_test.values, predicted_gru)
    print("Accuracy of GRU: {:.2f}%".format(gru_accuracy * 100))

Accuracy of GRU: 5.39%


## c) Long Short-Term Memory (LSTM)

In [155]:
# Define the LSTM model
class LSTMModel(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(LSTMModel, self).__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        lstm_out, _ = self.lstm(x)
        output = self.fc(lstm_out[:, -1, :])  # Use the last time step's output for classification
        return output




In [156]:
# Instantiate the LSTM model
lstm_model = LSTMModel(input_size=10, hidden_size=10, output_size=5)

# Define loss function and optimizer for LSTM
optimizer_lstm = optim.Adam(lstm_model.parameters(), lr=0.001)


In [157]:
# Training the LSTM model
for epoch in range(num_epochs):
    optimizer_lstm.zero_grad()
    lstm_outputs = lstm_model(X_train_word2vec_first10_tensor)
    loss_lstm = criterion(lstm_outputs, y_train_tensor)
    loss_lstm.backward()
    optimizer_lstm.step()

# Test the LSTM model
with torch.no_grad():
    test_lstm_outputs = lstm_model(X_test_word2vec_first10_tensor)
    _, predicted_lstm = torch.max(test_lstm_outputs, 1)
    lstm_accuracy = accuracy_score(y_test.values, predicted_lstm)
    print("Accuracy of LSTM: {:.2f}%".format(lstm_accuracy * 100))

Accuracy of LSTM: 8.82%


# Conclusion

In the evaluation of different recurrent neural network architectures, including Simple RNN, GRU, and LSTM, the obtained accuracies were 8.17%, 5.39%, and 8.82%, respectively. Despite these efforts, the results indicate a need for more sophisticated model architectures and hyperparameter tuning. Techniques like increasing network depth, experimenting with different activation functions, and adjusting dropout rates could enhance the models' performance. Additionally, integrating pre-trained embeddings and exploring transfer learning methods might provide valuable insights. Continuous iteration and exploration of advanced techniques are crucial to achieving higher accuracy and improving the models' overall predictive power in natural language processing tasks.